File: package/umt_rust/tests/string/format_string/test_get_value.rs
use umt_rust::obj;
use umt_rust::object::Value;
use umt_rust::string::format_string::get_value;

// Simple property access tests
#[test]
fn test_retrieve_simple_string_properties() {
    let obj = obj!("name" => "Alice", "age" => 30);

    assert_eq!(get_value(&obj, "name"), Some(&Value::String("Alice".to_string())));
    assert_eq!(get_value(&obj, "age"), Some(&Value::Int(30)));
}

#[test]
fn test_return_none_for_non_existent_properties() {
    let obj = obj!("name" => "Alice");

    assert_eq!(get_value(&obj, "nonexistent"), None);
    assert_eq!(get_value(&obj, "missing"), None);
}

#[test]
fn test_handle_various_data_types() {
    let obj = obj!(
        "string" => "text",
        "number" => 42,
        "boolean" => true,
        "nullValue" => Value::Null,
        "zero" => 0,
        "emptyString" => ""
    );

    assert_eq!(get_value(&obj, "string"), Some(&Value::String("text".to_string())));
    assert_eq!(get_value(&obj, "number"), Some(&Value::Int(42)));
    assert_eq!(get_value(&obj, "boolean"), Some(&Value::Bool(true)));
    assert_eq!(get_value(&obj, "nullValue"), Some(&Value::Null));
    assert_eq!(get_value(&obj, "zero"), Some(&Value::Int(0)));
    assert_eq!(get_value(&obj, "emptyString"), Some(&Value::String("".to_string())));
}

// Nested property access tests
#[test]
fn test_retrieve_nested_properties() {
    let obj = obj!(
        "user" => obj!(
            "name" => "Bob",
            "profile" => obj!(
                "age" => 25,
                "location" => "Tokyo"
            )
        )
    );

    assert_eq!(get_value(&obj, "user.name"), Some(&Value::String("Bob".to_string())));
    assert_eq!(get_value(&obj, "user.profile.age"), Some(&Value::Int(25)));
    assert_eq!(
        get_value(&obj, "user.profile.location"),
        Some(&Value::String("Tokyo".to_string()))
    );
}

#[test]
fn test_return_none_for_broken_nested_paths() {
    let obj = obj!(
        "user" => obj!(
            "name" => "Bob"
        )
    );

    assert_eq!(get_value(&obj, "user.nonexistent"), None);
    assert_eq!(get_value(&obj, "user.profile.age"), None);
    assert_eq!(get_value(&obj, "nonexistent.property"), None);
}

#[test]
fn test_handle_deep_nesting() {
    let obj = obj!(
        "level1" => obj!(
            "level2" => obj!(
                "level3" => obj!(
                    "level4" => obj!(
                        "value" => "deep"
                    )
                )
            )
        )
    );

    assert_eq!(
        get_value(&obj, "level1.level2.level3.level4.value"),
        Some(&Value::String("deep".to_string()))
    );
}

// Array access tests
#[test]
fn test_access_array_elements_with_positive_indices() {
    let obj = obj!(
        "items" => vec!["A", "B", "C", "D"]
    );

    assert_eq!(get_value(&obj, "items[0]"), Some(&Value::String("A".to_string())));
    assert_eq!(get_value(&obj, "items[1]"), Some(&Value::String("B".to_string())));
    assert_eq!(get_value(&obj, "items[2]"), Some(&Value::String("C".to_string())));
    assert_eq!(get_value(&obj, "items[3]"), Some(&Value::String("D".to_string())));
}

#[test]
fn test_access_array_elements_with_negative_indices() {
    let obj = obj!(
        "items" => vec!["A", "B", "C", "D"]
    );

    assert_eq!(get_value(&obj, "items[-1]"), Some(&Value::String("D".to_string())));
    assert_eq!(get_value(&obj, "items[-2]"), Some(&Value::String("C".to_string())));
    assert_eq!(get_value(&obj, "items[-3]"), Some(&Value::String("B".to_string())));
    assert_eq!(get_value(&obj, "items[-4]"), Some(&Value::String("A".to_string())));
}

#[test]
fn test_return_none_for_out_of_bounds_indices() {
    let obj = obj!(
        "items" => vec!["A", "B"]
    );

    assert_eq!(get_value(&obj, "items[5]"), None);
    assert_eq!(get_value(&obj, "items[-5]"), None);
}

#[test]
fn test_handle_empty_arrays() {
    let obj = obj!(
        "items" => Vec::<String>::new()
    );

    assert_eq!(get_value(&obj, "items[0]"), None);
    assert_eq!(get_value(&obj, "items[-1]"), None);
}

// Complex nested access tests
#[test]
fn test_handle_arrays_of_objects() {
    let obj = obj!(
        "users" => vec![
            obj!("name" => "Alice", "age" => 30),
            obj!("name" => "Bob", "age" => 25),
            obj!("name" => "Charlie", "age" => 35)
        ]
    );

    assert_eq!(get_value(&obj, "users[0].name"), Some(&Value::String("Alice".to_string())));
    assert_eq!(get_value(&obj, "users[1].age"), Some(&Value::Int(25)));
    assert_eq!(get_value(&obj, "users[-1].name"), Some(&Value::String("Charlie".to_string())));
}

#[test]
fn test_handle_nested_arrays() {
    let obj = obj!(
        "matrix" => vec![
            vec![1, 2, 3],
            vec![4, 5, 6],
            vec![7, 8, 9]
        ]
    );

    match get_value(&obj, "matrix[0]") {
        Some(Value::Array(arr)) => {
            assert_eq!(arr.len(), 3);
            assert_eq!(arr[0], Value::Int(1));
        },
        _ => panic!("Expected array"),
    }
}

#[test]
fn test_handle_objects_containing_arrays_containing_objects() {
    let obj = obj!(
        "data" => obj!(
            "categories" => vec![
                obj!(
                    "name" => "Technology",
                    "items" => vec![
                        obj!("title" => "Laptop", "price" => 1000),
                        obj!("title" => "Phone", "price" => 500)
                    ]
                ),
                obj!(
                    "name" => "Books",
                    "items" => vec![
                        obj!("title" => "Novel", "price" => 20)
                    ]
                )
            ]
        )
    );

    assert_eq!(
        get_value(&obj, "data.categories[0].name"),
        Some(&Value::String("Technology".to_string()))
    );
    assert_eq!(
        get_value(&obj, "data.categories[0].items[1].title"),
        Some(&Value::String("Phone".to_string()))
    );
    assert_eq!(
        get_value(&obj, "data.categories[-1].items[0].price"),
        Some(&Value::Int(20))
    );
}

// Edge cases tests
#[test]
fn test_handle_null_and_undefined_objects() {
    assert_eq!(get_value(&Value::Null, "property"), None);
}

#[test]
fn test_handle_primitive_values_as_objects() {
    assert_eq!(get_value(&Value::Int(42), "property"), None);
    assert_eq!(get_value(&Value::String("string".to_string()), "property"), None);
    assert_eq!(get_value(&Value::Bool(true), "property"), None);
}

#[test]
fn test_handle_empty_path() {
    let obj = obj!("name" => "Alice");
    assert_eq!(get_value(&obj, ""), None);
}

#[test]
fn test_handle_array_access_on_non_arrays() {
    let obj = obj!(
        "notAnArray" => "string"
    );

    assert_eq!(get_value(&obj, "notAnArray[0]"), None);
}

// Special path formats tests
#[test]
fn test_handle_zero_indices() {
    let obj = obj!(
        "items" => vec!["zero", "one", "two"]
    );

    assert_eq!(get_value(&obj, "items[0]"), Some(&Value::String("zero".to_string())));
    assert_eq!(get_value(&obj, "items[-0]"), Some(&Value::String("zero".to_string())));
}

#[test]
fn test_handle_array_indices_in_the_middle_of_paths() {
    let obj = obj!(
        "groups" => vec![
            obj!(
                "name" => "Group1",
                "subGroups" => vec![
                    obj!("name" => "SubGroup1", "value" => "test1"),
                    obj!("name" => "SubGroup2", "value" => "test2")
                ]
            )
        ]
    );

    assert_eq!(
        get_value(&obj, "groups[0].subGroups[1].value"),
        Some(&Value::String("test2".to_string()))
    );
}

// Data type preservation tests
#[test]
fn test_preserve_original_data_types() {
    let obj = obj!(
        "numbers" => vec![Value::Int(0), Value::Int(1), Value::Int(-1), Value::Float(3.14)],
        "booleans" => vec![true, false],
        "objects" => vec![obj!("key" => "value")],
        "mixed" => vec![Value::Null, Value::String("".to_string()), Value::Int(0), Value::Bool(false)]
    );

    assert_eq!(get_value(&obj, "numbers[0]"), Some(&Value::Int(0)));
    assert_eq!(get_value(&obj, "numbers[2]"), Some(&Value::Int(-1)));
    assert_eq!(get_value(&obj, "numbers[3]"), Some(&Value::Float(3.14)));
    assert_eq!(get_value(&obj, "booleans[0]"), Some(&Value::Bool(true)));
    assert_eq!(get_value(&obj, "booleans[1]"), Some(&Value::Bool(false)));

    // For objects, we check by matching
    let obj0 = get_value(&obj, "objects[0]").unwrap();
    if let Value::Object(map) = obj0 {
        assert_eq!(map.get("key"), Some(&Value::String("value".to_string())));
    } else {
        panic!("Expected object");
    }

    assert_eq!(get_value(&obj, "mixed[0]"), Some(&Value::Null));
    assert_eq!(get_value(&obj, "mixed[1]"), Some(&Value::String("".to_string())));
    assert_eq!(get_value(&obj, "mixed[2]"), Some(&Value::Int(0)));
    assert_eq!(get_value(&obj, "mixed[3]"), Some(&Value::Bool(false)));
}

File: package/umt_python/tests/unit/string/test_get_value.py
import unittest

from src.string import get_value


class TestGetValueSimple(unittest.TestCase):
    def test_simple_properties(self):
        obj = {"name": "Alice", "age": 30}
        self.assertEqual(get_value(obj, "name"), "Alice")
        self.assertEqual(get_value(obj, "age"), 30)

    def test_non_existent_properties(self):
        obj = {"name": "Alice"}
        self.assertIsNone(get_value(obj, "nonexistent"))
        self.assertIsNone(get_value(obj, "missing"))

    def test_various_data_types(self):
        obj = {
            "string": "text",
            "number": 42,
            "boolean": True,
            "nullValue": None,
            "zero": 0,
            "emptyString": "",
        }
        self.assertEqual(get_value(obj, "string"), "text")
        self.assertEqual(get_value(obj, "number"), 42)
        self.assertEqual(get_value(obj, "boolean"), True)
        self.assertIsNone(get_value(obj, "nullValue"))
        self.assertEqual(get_value(obj, "zero"), 0)
        self.assertEqual(get_value(obj, "emptyString"), "")


class TestGetValueNested(unittest.TestCase):
    def test_nested_properties(self):
        obj = {"user": {"name": "Bob", "profile": {"age": 25, "location": "Tokyo"}}}
        self.assertEqual(get_value(obj, "user.name"), "Bob")
        self.assertEqual(get_value(obj, "user.profile.age"), 25)
        self.assertEqual(get_value(obj, "user.profile.location"), "Tokyo")

    def test_broken_nested_paths(self):
        obj = {"user": {"name": "Bob"}}
        self.assertIsNone(get_value(obj, "user.nonexistent"))
        self.assertIsNone(get_value(obj, "user.profile.age"))
        self.assertIsNone(get_value(obj, "nonexistent.property"))

    def test_deep_nesting(self):
        obj = {"level1": {"level2": {"level3": {"level4": {"value": "deep"}}}}}
        self.assertEqual(get_value(obj, "level1.level2.level3.level4.value"), "deep")


class TestGetValueArray(unittest.TestCase):
    def test_positive_indices(self):
        obj = {"items": ["A", "B", "C", "D"]}
        self.assertEqual(get_value(obj, "items[0]"), "A")
        self.assertEqual(get_value(obj, "items[1]"), "B")
        self.assertEqual(get_value(obj, "items[2]"), "C")
        self.assertEqual(get_value(obj, "items[3]"), "D")

    def test_negative_indices(self):
        obj = {"items": ["A", "B", "C", "D"]}
        self.assertEqual(get_value(obj, "items[-1]"), "D")
        self.assertEqual(get_value(obj, "items[-2]"), "C")
        self.assertEqual(get_value(obj, "items[-3]"), "B")
        self.assertEqual(get_value(obj, "items[-4]"), "A")

    def test_out_of_bounds(self):
        obj = {"items": ["A", "B"]}
        self.assertIsNone(get_value(obj, "items[5]"))
        self.assertIsNone(get_value(obj, "items[-5]"))

    def test_empty_arrays(self):
        obj = {"items": []}
        self.assertIsNone(get_value(obj, "items[0]"))
        self.assertIsNone(get_value(obj, "items[-1]"))


class TestGetValueComplex(unittest.TestCase):
    def test_arrays_of_objects(self):
        obj = {
            "users": [
                {"name": "Alice", "age": 30},
                {"name": "Bob", "age": 25},
                {"name": "Charlie", "age": 35},
            ]
        }
        self.assertEqual(get_value(obj, "users[0].name"), "Alice")
        self.assertEqual(get_value(obj, "users[1].age"), 25)
        self.assertEqual(get_value(obj, "users[-1].name"), "Charlie")

    def test_nested_arrays(self):
        obj = {"matrix": [[1, 2, 3], [4, 5, 6], [7, 8, 9]]}
        self.assertEqual(get_value(obj, "matrix[0]"), [1, 2, 3])
        self.assertEqual(get_value(obj, "matrix[1]"), [4, 5, 6])
        self.assertEqual(get_value(obj, "matrix[-1]"), [7, 8, 9])

    def test_objects_containing_arrays_containing_objects(self):
        obj = {
            "data": {
                "categories": [
                    {
                        "name": "Technology",
                        "items": [
                            {"title": "Laptop", "price": 1000},
                            {"title": "Phone", "price": 500},
                        ],
                    },
                    {"name": "Books", "items": [{"title": "Novel", "price": 20}]},
                ]
            }
        }
        self.assertEqual(get_value(obj, "data.categories[0].name"), "Technology")
        self.assertEqual(get_value(obj, "data.categories[0].items[1].title"), "Phone")
        self.assertEqual(get_value(obj, "data.categories[-1].items[0].price"), 20)


class TestGetValueEdgeCases(unittest.TestCase):
    def test_null_and_undefined_objects(self):
        self.assertIsNone(get_value(None, "property"))

    def test_primitive_values_as_objects(self):
        self.assertIsNone(get_value(42, "property"))
        self.assertIsNone(get_value("string", "property"))
        self.assertIsNone(get_value(True, "property"))

    def test_empty_path(self):
        obj = {"name": "Alice"}
        self.assertIsNone(get_value(obj, ""))

    def test_array_access_on_non_arrays(self):
        obj = {"notAnArray": "string"}
        self.assertIsNone(get_value(obj, "notAnArray[0]"))

    def test_complex_array_notation(self):
        obj = {"complex[key]": "value", "items": ["A", "B"]}
        self.assertEqual(get_value(obj, "items[0]"), "A")
        self.assertEqual(get_value(obj, "complex[key]"), "value")


class TestGetValueSpecialPaths(unittest.TestCase):
    def test_zero_indices(self):
        obj = {"items": ["zero", "one", "two"]}
        self.assertEqual(get_value(obj, "items[0]"), "zero")
        self.assertEqual(get_value(obj, "items[-0]"), "zero")

    def test_array_indices_in_middle_of_paths(self):
        obj = {
            "groups": [
                {
                    "name": "Group1",
                    "subGroups": [
                        {"name": "SubGroup1", "value": "test1"},
                        {"name": "SubGroup2", "value": "test2"},
                    ],
                }
            ]
        }
        self.assertEqual(get_value(obj, "groups[0].subGroups[1].value"), "test2")


class TestGetValuePrototypePollution(unittest.TestCase):
    def test_block_proto_key_access(self):
        obj = {"a": {"b": 1}}
        self.assertIsNone(get_value(obj, "__proto__"))
        self.assertIsNone(get_value(obj, "__proto__.polluted"))
        self.assertIsNone(get_value(obj, "a.__proto__"))

    def test_block_constructor_key_access(self):
        obj = {"a": {"b": 1}}
        self.assertIsNone(get_value(obj, "constructor"))
        self.assertIsNone(get_value(obj, "constructor.prototype"))
        self.assertIsNone(get_value(obj, "a.constructor"))

    def test_allow_normal_non_dangerous_keys(self):
        obj = {"proto": "safe", "construct": "ok", "proto_type": "fine"}
        self.assertEqual(get_value(obj, "proto"), "safe")
        self.assertEqual(get_value(obj, "construct"), "ok")
        self.assertEqual(get_value(obj, "proto_type"), "fine")


class TestGetValueDataTypes(unittest.TestCase):
    def test_preserve_original_data_types(self):
        obj = {
            "numbers": [0, 1, -1, 3.14],
            "booleans": [True, False],
            "objects": [{"key": "value"}],
            "mixed": [None, "", 0, False],
        }
        self.assertEqual(get_value(obj, "numbers[0]"), 0)
        self.assertEqual(get_value(obj, "numbers[2]"), -1)
        self.assertEqual(get_value(obj, "numbers[3]"), 3.14)
        self.assertEqual(get_value(obj, "booleans[0]"), True)
        self.assertEqual(get_value(obj, "booleans[1]"), False)
        self.assertEqual(get_value(obj, "objects[0]"), {"key": "value"})
        self.assertIsNone(get_value(obj, "mixed[0]"))
        self.assertEqual(get_value(obj, "mixed[1]"), "")
        self.assertEqual(get_value(obj, "mixed[2]"), 0)
        self.assertEqual(get_value(obj, "mixed[3]"), False)


if __name__ == "__main__":
    unittest.main()
